gluonnlp/data/candidate_sampler.py

```diff
@@ -83,12 +108,36 @@ def __init__(self, weights):
         self.prob = mx.nd.array(self.prob)
         self.alias = mx.nd.array(self.alias)
 
-    def __call__(self, k):
-        """Draw k samples from the distribution."""
-        idx = mx.nd.array(np.random.randint(0, self.N, size=k))
+    def __call__(self, shape, true_samples=None, true_samples_mask=None):
+        """Draw shape samples from the distribution.
+
+        If true_samples is specified, also returns a mask that masks any random
+        elements that are also part of the same row as true_samples.
+
+        """
+
+        # Draw samples
+        idx = mx.nd.array(np.random.randint(0, self.N, size=shape))
         prob = self.prob[idx]
         alias = self.alias[idx]
-        where = mx.nd.random.uniform(shape=k) < prob
+        where = mx.nd.random.uniform(shape=shape) < prob
         hit = idx * where
         alt = alias * (1 - where)
-        return hit + alt
+        candidates = hit + alt
+
+        # Remove accidental hits
+        if true_samples is not None:
+            candidates_np = candidates.asnumpy()
```

Comment:
The conversion to  numpy is necessary for using just in time compilation with numba for the `_candidates_mask` function. Alternatively the negative samples could be shared among all words in a batch (as commonly done for language modeling https://arxiv.org/abs/1602.02410 http://www.aclweb.org/anthology/N16-1145.pdf), allowing the use of a simple broadcast comparision as in https://github.com/apache/incubator-mxnet/blob/master/example/rnn/large_word_lm/model.py#L115-L118 .  On the other hand this would implicitly change the sampling distribution as unrelated words (ie. words that seldomly co-occur in a context) would be more likely masked (given that they may still occur in the same batch)